HIRS_Utils/src/main/java/hirs/tpm/eventlog/uefi/UefiDevicePath.java

```diff
@@ -0,0 +1,441 @@
+ *
+ * #define EFI_DEVICE_PATH_PROTOCOL_GUID \09576e91-6d3f-11d2-8e39-00a0c969723b
+ *  typedef struct _EFI_DEVICE_PATH_PROTOCOL {
+ *      UINT8 Type;
+ *      UINT8 SubType;
+ *      UINT8 Length[2];
+ *    } EFI_DEVICE_PATH_PROTOCOL;
+ *
+ *  Where Type is defined in the UEFI spec section 10:
+ *    Type 0x01  Hardware Device Path
+ *    Type 0x02  ACPI Device Path
+ *    Type 0x03  Messaging Device Path
+ *    Type 0x04  Media Device Path
+ *    Type 0x05  BIOS Boot Specification Device Path
+ *    Type 0x7F  End of Hardware Device Path
+ *  Each Type has a sub-type that may or may no be defined in the section
+ *
+ *  Only a few of the SubTypes have been implemented as there are many,
+ *     but only those that were reported using the test devices at hand.
+ *     Without test patterns, the processing may lead to an un-handled exception
+ */
+public class UefiDevicePath {
+    /** UEFI Device path type. */
+    private String type = "";
+    /** UEFI Device path sub-type. */
+    private String subType = "";
+    /** UEFI Device path human readable description. */
+    private String devPathInfo = "";
+    /** UEFI Device path length. */
+    private int length = 0;
+
+/**
+ * UEFI Device path constructor.
+ * @param path byte array holding device path data
+ * @throws UnsupportedEncodingException if path byte array contains unexpected values
+ */
+ public UefiDevicePath(final byte[] path) throws UnsupportedEncodingException {
+   devPathInfo = processDevPath(path);
+   byte[] lengthBytes = new byte[UefiConstants.SIZE_2];
+   System.arraycopy(path, UefiConstants.OFFSET_2, lengthBytes, 0, UefiConstants.OFFSET_2);
+   length = HexUtils.leReverseInt(lengthBytes);
+   }
+
+/**
+ * Returns the UEFI device type.
+ * @return uefi type
+ */
+ public String getType() {
+   return type;
+ }
+
+/**
+ * Returns the UEFI device sub-type.
+ * @return uefi sub-type
+ */
+ public String getSubType() {
+   return subType.trim();
+ }
+
+ /**
+  * Returns the UEFI device structure length.
+  * @return uefi device structure length
+  */
+ public int getLegth() {
+  return length;
+ }
+
+/**
+ * Processes the UEFI device path.
+ * UEFI device path is a collection of EFI_DEVICE_PATH_PROTOCOL structures of variable length.
+ * length must be calculated for each device path and used as an offset.
+ * devPath is terminated by 07f and 0xff per the UEFi spec.
+ * @param path byte array holding the Device path
+ * @return Human readable string containing the device path description.
+ * @throws UnsupportedEncodingException
+ */
+ private String processDevPath(final byte[] path) throws UnsupportedEncodingException {
+  StringBuffer pInfo = new StringBuffer();
+  String devicePathInfo = "";
+  int devLength = 0, pathOffset = 0;
+  boolean moreDev = true;
+  while (moreDev) {
+  Byte devPath = Byte.valueOf(path[pathOffset]);
+  if ((devPath.intValue() == UefiConstants.TERMINATOR)
+          || (devPath.intValue() == UefiConstants.END_FLAG)) {
+    moreDev = false;
```

Comment:
This assignment is never used because the break immediately after does what the assignment is meant to do.  It would be better to skip creating the variable at all, assigning the While loop to true and just use breaks where it should stop.  or remove the break;  However using break is better since you stop execution and don't have to process/execute anymore code that is not needed if the purpose is to stop the loop; in which break should replace all moreDev = false statements.

Reply:
OK.